homeassistant/components/bmw_connected_drive/quality_scale.yaml

```diff
@@ -0,0 +1,80 @@
+rules:
+  # Bronze
+  action-setup:
+    status: exempt
+    comment: |
+      Does not have custom services
+  appropriate-polling: done
+  brands: done
+  common-modules: done
+  config-flow-test-coverage: done
+  config-flow: done
+  dependency-transparency: done
+  docs-actions:
+    status: exempt
+    comment: |
+      Does not have custom services
+  docs-high-level-description: done
+  docs-installation-instructions: done
+  docs-removal-instructions: todo
+  entity-event-setup:
+    status: exempt
+    comment: |
+      This integration doesn't have any events.
+  entity-unique-id: done
+  has-entity-name: done
+  runtime-data: done
+  test-before-configure: done
+  test-before-setup: done
+  unique-config-entry: done
+
+  # Silver
+  action-exceptions:
+    status: exempt
+    comment: |
+      Does not have custom services
+  config-entry-unloading: done
+  docs-configuration-parameters: done
+  docs-installation-parameters: done
+  entity-unavailable: done
+  integration-owner: done
+  log-when-unavailable: done
+  parallel-updates: done
+  reauthentication-flow: done
+  test-coverage: done
```

Comment:
Out of curiosity, why do we care about significant state here?
```py
    # wait for the recorder to really store the data
    await async_wait_recording_done(hass)
    states = await hass.async_add_executor_job(
        get_significant_states, hass, now, None, [entity_id]
    )
    assert any(s for s in states[entity_id] if s.state == STATE_UNKNOWN) is False
```

Reply:
We change the lock state when the user presses a button (ahead of time). If the command execution fails (depending on the vehicle 10s up to 1 minute), it'll be set to `UNKOWN`, then the integrations refreshed all data and sets it again to the actual value supplied by the API.

`test_service_call_success` checks that no `UNKOWN` state was stored, while `test_service_call_fail` verifies exactly this.

Reply:
When calling the service, the integrations sets the target value optimistically (https://github.com/home-assistant/core/blob/dev/homeassistant/components/bmw_connected_drive/lock.py#L63-L66 and https://github.com/home-assistant/core/blob/dev/homeassistant/components/bmw_connected_drive/lock.py#L87-L90).
I'm using the state machine to verify that an `UNKOWN` state is set.

Start with `LOCKED` > `UNLOCKED` (optimistic) > `UNKNOWN` (failed call) > `LOCKED`. As this all happens during runtime of the blocking service call, I'm using `get_significant_states()` to check this.

If you have another idea how to test this or think it could be removed, please let me know.